### PR TITLE
[ty] Extend test coverage for parsing of reStructuredText literal blocks

### PR DESCRIPTION
**Pull request by lerebear**
## Summary

This adds some additional test coverage that demonstrates that we parse certain edge cases involving reStructuredText (reST) literal blocks correctly.

In particular, I learned that continuation lines and nest body elements of a reST field body [**must** be indented relative to the field name marker](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#field-lists:~:text=The%20field%20body%20may%20contain%20multiple%20body%20elements%2C%20indented%20relative%20to%20the%20field%20marker.%20The%20first%20line%20after%20the%20field%20name%20marker%20determines%20the%20indentation%20of%20the%20field%20body.). So although a [quoted literal block](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#quoted-literal-blocks), does not require additional indentation if it appears on its own, it must have indentation relative to the field name marker if it is part of a field body.

The current parser (incidentally) handles this correctly; this helps to ensure that we don't regress that behaviour.

## Test Plan

Please see included tests.

### CODE REVIEW

File: crates/ty_ide/src/docstring/formats/rst.rs [hunk 1]
```diff
@@ -72,6 +72,18 @@ impl<'a> Lines<'a> {
         let (index, line) = self.inner.next()?;
         Some(DocstringLine::new(index, &line))
     }
+
+    /// Returns the next non-blank line without advancing the cursor.
+    fn peek_non_blank(&self) -> Option<DocstringLine<'a>> {
+        let mut next = self.clone();
```
Comment (MichaReiser) [inline crates/ty_ide/src/docstring/formats/rst.rs]: Nit: Maybe `lookahead` or `lines`? The iterator itself isn't the `next` line

You could also use `.next()` directly, since you don't need to be concerned about not progressing the iterator (you work on a clone)

```rust
    std::iter::from_fn(|| next.next())
        .find(|line| !line.text.trim().is_empty())
```

File: crates/ty_ide/src/docstring/preformatted.rs [hunk 2]
```diff
@@ -90,54 +90,76 @@ impl RestLiteralBlockScanner {
...
+    pub(super) fn would_consume_line(&self, line: &str) -> bool {
+        self.line_outcome(line).is_literal_block
+    }
+
+    fn line_outcome(&self, line: &str) -> RestLiteralBlockLineOutcome {
         let current_indent = indentation(line);
         let line_is_empty = line.trim_start().is_empty();
 
-        match self.state {
+        match &self.state {
```
Comment (MichaReiser) [inline crates/ty_ide/src/docstring/preformatted.rs]: `state` is copy, and it helps remove many of the somewhat awkward `*`]

```suggestion
        match self.state {
```

**Review by MichaReiser — APPROVED** [review]

File: crates/ty_ide/src/docstring/preformatted.rs [hunk 3]
```diff
@@ -90,54 +90,76 @@ impl RestLiteralBlockScanner {
...
+        if let Some(next_state) = outcome.next_state {
+            self.state = next_state;
+        }
+
+        outcome.is_literal_block
+    }
+
+    /// Returns whether `consume_line` would consume the given line.
+    pub(super) fn would_consume_line(&self, line: &str) -> bool {
+        self.line_outcome(line).is_literal_block
```
Comment (MichaReiser) [inline crates/ty_ide/src/docstring/preformatted.rs]: Nit: You could consider calling `self.clone().consume_line` here, to avoid the new `line_outcome` abstraction. Assuming this isn't called too frequently

**Comment by MichaReiser** [comment]
Codex found a case where the literal parsing now applies too aggressively. What I understand the relevant part of the spec is 

> All whitespace (including line breaks, but excluding minimum indentation for indented literal blocks) is preserved. Blank lines are required before and after a literal block, but these blank lines are not included as part of the literal block.


regresssion:

```
:param first: Explanation::
:param second: Second parameter.
```

The `Explanation::` lacks the blank line to introduce a literal block. So the expected output is to parse `second` as a field


```
first: Explanation::
second: Second parameter.
```

Instead, it's parsed as a literal block

```
first: Explanation::
  :param second: Second parameter.
```

File: crates/ty_ide/src/docstring/formats/rst.rs (hunk 1 above)
Comment (lerebear) [reply crates/ty_ide/src/docstring/formats/rst.rs]: Fixed in latest push.

File: crates/ty_ide/src/docstring/preformatted.rs (hunk 2 above)
Comment (lerebear) [reply crates/ty_ide/src/docstring/preformatted.rs]: Thanks: fixed in latest push.

File: crates/ty_ide/src/docstring/preformatted.rs (hunk 3 above)
Comment (lerebear) [reply crates/ty_ide/src/docstring/preformatted.rs]: Thanks, this was great simplification!

**Comment by lerebear** [comment]
> Codex found a case where the literal parsing now applies too aggressively

Fixed in latest push.

**Comment by lerebear** [comment]
@MichaReiser, FYI that I changed the intent of this PR completely because during my final review I realized that I had misinterpreted the reST spec. As such, this just expands test coverage to lock in the correct behaviour.
